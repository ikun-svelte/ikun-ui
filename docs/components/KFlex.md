---
title: KFlex
lang: en-US
---

# KFlex

flex layout component

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/flex
```

```bash [yarn]
yarn add @ikun-ui/flex
```

```bash [npm]
npm install @ikun-ui/flex
```

:::

## Basic usage

The basic usage.

<demo src="flex/basic.svelte"  github='Flex'></demo>

## Justify and align

Supports adjusting justify and align,
the values should conform to the [justify <span class="i-carbon-link text-12px" />](https://www.tailwindcss.cn/docs/justify-content)
and [align <span class="i-carbon-link text-12px" />](https://www.tailwindcss.cn/docs/align-items) values of tailwindcss

<demo src="flex/justify-align.svelte" github='Flex'></demo>

## Gap

The gap attribute can jump the gap and supports `sm`, `md`, `lg` and custom numbers.

<demo src="flex/gap.svelte" github='Flex'></demo>

## Wrap

Auto wrap line.

<demo src="flex/wrap.svelte" github='Flex'></demo>

## Flex Props

| Name     | Type                                                                                                  | Default                                                           | Description                                                                 |
| -------- | ----------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------- | --------------------------------------------------------------------------- |
| vertical | `boolean`                                                                                             | `false`                                                           | Is direction of the flex vertical.                                          |
| wrap     | `'wrap' \| 'nowrap' \| 'wrap-reverse'`                                                                | `'normal'`                                                        | Set whether the element is displayed in a single line or in multiple lines. |
| justify  | `'normal' \| 'start'\| 'end'\| 'center'\| 'between'\| 'around' \| 'evenly' \| 'stretch'` | `'normal'` | Sets the alignment of elements in the direction of the main axis. |
| align    | `'center' \| 'start' \| 'end' \| 'baseline' \| 'normal' \| 'stretch'`                                 | `'normal'`                                                        | Sets the alignment of elements in the direction of the cross axis.          |
| flex     | `string`                                                                                              | `-`                                                               | Flex CSS shorthand properties.                                              |
| gap      | `'sm' \| 'md' \| 'lg' \| number`                                                                      | `-`                                                               | Sets the gap between grids.                                                 |
| elm      | `string`                                                                                              | `div`                                                             | Custom element type.                                                        |
| cls      | `string`                                                                                              | `''`                                                              | Additional class for component                                              |
| attrs    | `any`                                                                                                 | `{}`                                                              | Additional attributes                                                       |

## Flex Slots

| Name    | Description                    |
| ------- | ------------------------------ |
| default | content of the flex component. |
