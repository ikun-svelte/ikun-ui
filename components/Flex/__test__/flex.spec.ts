import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import KFlex from '../src';
import KFlexWrap from './fixture/wrap.svelte';
import KFlexJustify from './fixture/justify.svelte';
import KFlexAlign from './fixture/align.svelte';
import KFlexGap from './fixture/gap.svelte';
let host;

const initHost = () => {
	host = globalThis.document.createElement('div');
	host.setAttribute('id', 'host');
	globalThis.document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
	vi.useFakeTimers();
});
afterEach(() => {
	host.remove();
	vi.useRealTimers();
});

describe('Test: KFlex', () => {
	test('props: cls', async () => {
		//@ts-ignore
		const instance = new KFlex({
			target: host,
			props: {
				cls: 'k-flex--test'
			}
		});
		expect(instance).toBeTruthy();
		expect(host!.innerHTML.includes('k-flex--test')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: vertical', async () => {
		//@ts-ignore
		const instance = new KFlex({
			target: host,
			props: {
				vertical: true
			}
		});
		expect(instance).toBeTruthy();
		expect(host!.innerHTML.includes('k-flex--col')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: flex', async () => {
		//@ts-ignore
		const instance = new KFlex({
			target: host,
			props: {
				flex: 1
			}
		});
		expect(instance).toBeTruthy();
		expect(host!.innerHTML.includes('style="flex: 1;"')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: elm', async () => {
		//@ts-ignore
		const instance = new KFlex({
			target: host,
			props: {
				elm: 'p'
			}
		});
		expect(instance).toBeTruthy();
		expect(host!.innerHTML.includes('</div>')).not.toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: wrap', async () => {
		//@ts-ignore
		const instance = new KFlexWrap({
			target: host
		});
		expect(instance).toBeTruthy();
		const flexElms = host.querySelectorAll('.k-flex');
		expect(flexElms[0].className.includes('k-flex--wrap-wrap')).toBeTruthy();
		expect(flexElms[1].className.includes('k-flex--wrap-nowrap')).toBeTruthy();
		expect(flexElms[2].className.includes('k-flex--wrap-wrap-reverse')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: justify', async () => {
		//@ts-ignore
		const instance = new KFlexJustify({
			target: host
		});
		expect(instance).toBeTruthy();
		const flexElms = host.querySelectorAll('.k-flex');
		expect(flexElms[0].className.includes('k-flex--justify-normal')).toBeTruthy();
		expect(flexElms[1].className.includes('k-flex--justify-start')).toBeTruthy();
		expect(flexElms[2].className.includes('k-flex--justify-end')).toBeTruthy();
		expect(flexElms[3].className.includes('k-flex--justify-center')).toBeTruthy();
		expect(flexElms[4].className.includes('k-flex--justify-between')).toBeTruthy();
		expect(flexElms[5].className.includes('k-flex--justify-around')).toBeTruthy();
		expect(flexElms[6].className.includes('k-flex--justify-evenly')).toBeTruthy();
		expect(flexElms[7].className.includes('k-flex--justify-stretch')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: align', async () => {
		//@ts-ignore
		const instance = new KFlexAlign({
			target: host
		});
		expect(instance).toBeTruthy();
		const flexElms = host.querySelectorAll('.k-flex');
		expect(flexElms[0].className.includes('k-flex--align-normal')).toBeTruthy();
		expect(flexElms[1].className.includes('k-flex--align-start')).toBeTruthy();
		expect(flexElms[2].className.includes('k-flex--align-end')).toBeTruthy();
		expect(flexElms[3].className.includes('k-flex--align-center')).toBeTruthy();
		expect(flexElms[4].className.includes('k-flex--align-baseline')).toBeTruthy();
		expect(flexElms[5].className.includes('k-flex--align-stretch')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: gap', async () => {
		//@ts-ignore
		const instance = new KFlexGap({
			target: host
		});
		expect(instance).toBeTruthy();
		const flexElms = host.querySelectorAll('.k-flex');
		expect(flexElms[0].style.gap).toBe('90px');
		expect(flexElms[1].style.gap).toBe('16px');
		expect(flexElms[2].style.gap).toBe('8px');
		expect(flexElms[3].style.gap).toBe('24px');
		expect(host.innerHTML).matchSnapshot();
	});
});
