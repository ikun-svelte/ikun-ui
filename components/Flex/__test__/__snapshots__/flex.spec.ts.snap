// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KFlex > props: align 1`] = `"<div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-normal k-flex--justify-normal"></div> <div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-start k-flex--justify-normal"></div> <div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-end k-flex--justify-normal"></div> <div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-center k-flex--justify-normal"></div> <div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-baseline k-flex--justify-normal"></div> <div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-stretch k-flex--justify-normal"></div>"`;

exports[`Test: KFlex > props: cls 1`] = `"<div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-normal k-flex--justify-normal k-flex--test"></div>"`;

exports[`Test: KFlex > props: elm 1`] = `"<p class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-normal k-flex--justify-normal"></p>"`;

exports[`Test: KFlex > props: flex 1`] = `"<div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-normal k-flex--justify-normal" style="flex: 1;"></div>"`;

exports[`Test: KFlex > props: gap 1`] = `"<div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-normal k-flex--justify-normal" style="gap: 90px;"></div> <div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-normal k-flex--justify-normal" style="gap: 16px;"></div> <div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-normal k-flex--justify-normal" style="gap: 8px;"></div> <div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-normal k-flex--justify-normal" style="gap: 24px;"></div>"`;

exports[`Test: KFlex > props: justify 1`] = `"<div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-normal k-flex--justify-normal"></div> <div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-normal k-flex--justify-start"></div> <div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-normal k-flex--justify-end"></div> <div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-normal k-flex--justify-center"></div> <div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-normal k-flex--justify-between"></div> <div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-normal k-flex--justify-around"></div> <div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-normal k-flex--justify-evenly"></div> <div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-normal k-flex--justify-stretch"></div>"`;

exports[`Test: KFlex > props: vertical 1`] = `"<div class="k-flex k-flex--col k-flex--wrap-nowrap k-flex--align-normal k-flex--justify-normal"></div>"`;

exports[`Test: KFlex > props: wrap 1`] = `"<div class="k-flex k-flex--row k-flex--wrap-wrap k-flex--align-normal k-flex--justify-normal"></div> <div class="k-flex k-flex--row k-flex--wrap-nowrap k-flex--align-normal k-flex--justify-normal"></div> <div class="k-flex k-flex--row k-flex--wrap-wrap-reverse k-flex--align-normal k-flex--justify-normal"></div>"`;
